cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/ProgramLifecycleHttpHandlerTest.java

```diff
@@ -309,6 +312,38 @@ public void testMapreduceHistory() throws Exception {
   }
 
   /**
+   * Tests history of a non existing program
+   */
+  @Test
+  public void testNonExistingProgramHistory() throws Exception {
+
+    Id.Program program = Id.Program.from(TEST_NAMESPACE2, DUMMY_APP_ID, ProgramType.MAPREDUCE, DUMMY_MR_NAME);
+    String namespace = program.getNamespaceId();
+    try {
+      deploy(DummyAppWithTrackingTable.class, Constants.Gateway.API_VERSION_3_TOKEN, namespace);
+      // first run
+      startProgram(program);
+      waitState(program, ProgramRunStatus.RUNNING.toString());
+      final String statusUrl = getVersionedAPIPath("apps/" + program.getApplicationId() +
+                                                     ProgramType.MAPREDUCE + "/NonExisting",
+                                                   Constants.Gateway.API_VERSION_3_TOKEN,
+                                                   program.getNamespaceId());
+      Assert.assertEquals(404, doPost(statusUrl).getStatusLine().getStatusCode());
+      stopProgram(program);
+      waitState(program, STOPPED);
+    } catch (Exception e) {
+      LOG.error("Got exception: ", e);
+    } finally {
+      HttpResponse httpResponse = doDelete(getVersionedAPIPath("apps/" + program.getApplicationId(),
+                                                               Constants.Gateway.API_VERSION_3_TOKEN,
+                                                               program.getNamespaceId()));
+      Assert.assertEquals(EntityUtils.toString(httpResponse.getEntity()),
```

Comment:
Assert shouldn't be in finally. The reason is that if there is failure in the `try` block, this `finally` will still get executed, which most likely will fail on the assert as well, which will shield the actual exception from the `try` block.